File: app/src/main/java/com/aexon/widget/AexonButton.java
package com.aexon.widget;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.Outline;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.RippleDrawable;
import android.os.Build;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.View;
import android.view.ViewOutlineProvider;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.aexon.R;

public class AexonButton extends Button {
	
	public static final int ICON_GRAVITY_START = 0x1;
	public static final int ICON_GRAVITY_TEXT_START = 0x2;
	public static final int ICON_GRAVITY_END = 0x3;
	public static final int ICON_GRAVITY_TEXT_END = 0x4;
	
	private float cornerRadius = 0f;
	private int iconGravity = ICON_GRAVITY_START;
	private float iconPadding = 0f;
	private int iconTint = Integer.MIN_VALUE;
	private int iconSize = 0;
	private int iconLeft = 0;
	
	@Nullable
	private Drawable icon = null;
	
	private int bgColor = Color.TRANSPARENT;
	private int rippleColor = Color.argb(40, 255, 255, 255);
	private int strokeColor = Color.TRANSPARENT;
	private float strokeWidth = 0f;
	
	public AexonButton(@NonNull Context context) {
		super(context);
		init(context, null);
	}
	
	public AexonButton(@NonNull Context context, @Nullable AttributeSet attrs) {
		super(context, attrs);
		init(context, attrs);
	}
	
	public AexonButton(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		init(context, attrs);
	}
	
	private void init(@NonNull Context context, @Nullable AttributeSet attrs) {
		setAllCaps(false);
		setMinimumWidth(dpToPx(context, 88f));
		setMinimumHeight(dpToPx(context, 36f));
		setIncludeFontPadding(false);
		setGravity(Gravity.CENTER);
		
		if (attrs != null) {
			TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.AexonButton);
			cornerRadius = a.getDimension(R.styleable.AexonButton_cornerRadius, 0f);
			iconGravity = a.getInt(R.styleable.AexonButton_iconGravity, ICON_GRAVITY_START);
			iconPadding = a.getDimension(R.styleable.AexonButton_iconPadding, dpToPx(context, 8f));
			iconTint = a.getColor(R.styleable.AexonButton_iconTint, Integer.MIN_VALUE);
			iconSize = (int) a.getDimension(R.styleable.AexonButton_iconSize, 0f);
			icon = a.getDrawable(R.styleable.AexonButton_ax_icon);
			bgColor = a.getColor(R.styleable.AexonButton_buttonColor, Color.TRANSPARENT);
			rippleColor = a.getColor(R.styleable.AexonButton_rippleColor, Color.argb(40, 255, 255, 255));
			strokeColor = a.getColor(R.styleable.AexonButton_strokeColor, Color.TRANSPARENT);
			strokeWidth = a.getDimension(R.styleable.AexonButton_strokeWidth, 0f);
			a.recycle();
		}
		
		if (bgColor == Color.TRANSPARENT) {
			Drawable bg = getBackground();
			if (bg instanceof ColorDrawable) {
				bgColor = ((ColorDrawable) bg).getColor();
			}
		}
		
		setCompoundDrawablePadding((int) iconPadding);
		applyBackground();
		updateIcon(true);
	}
	
	private void applyBackground() {
		GradientDrawable shape = new GradientDrawable();
		shape.setColor(bgColor);
		shape.setCornerRadius(cornerRadius);
		
		if (strokeWidth > 0 && strokeColor != Color.TRANSPARENT) {
			shape.setStroke((int) strokeWidth, strokeColor);
		}
		
		GradientDrawable mask = new GradientDrawable();
		mask.setShape(GradientDrawable.RECTANGLE);
		mask.setColor(Color.WHITE);
		mask.setCornerRadius(cornerRadius);
		
		RippleDrawable ripple = new RippleDrawable(ColorStateList.valueOf(rippleColor), shape, mask);
		setBackground(ripple);
		
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
			setOutlineProvider(new ViewOutlineProvider() {
				@Override
				public void getOutline(View view, Outline outline) {
					if (view.getWidth() > 0 && view.getHeight() > 0) {
						outline.setRoundRect(0, 0, view.getWidth(), view.getHeight(), cornerRadius);
					}
				}
			});
			setClipToOutline(true);
			invalidateOutline();
		}
	}
	
	private void updateIcon(boolean needsIconReset) {
		if (icon != null) {
			icon = icon.mutate();
			
			int width = iconSize != 0 ? iconSize : icon.getIntrinsicWidth();
			int height = iconSize != 0 ? iconSize : icon.getIntrinsicHeight();
			icon.setBounds(iconLeft, 0, iconLeft + width, height);
			
			if (iconTint != Integer.MIN_VALUE) {
				icon.setColorFilter(iconTint, PorterDuff.Mode.SRC_IN);
			} else {
				icon.clearColorFilter();
			}
		}
		
		if (needsIconReset) {
			resetIconDrawable();
		}
	}
	
	private void resetIconDrawable() {
		if (iconGravity == ICON_GRAVITY_START || iconGravity == ICON_GRAVITY_TEXT_START) {
			setCompoundDrawablesRelative(icon, null, null, null);
		} else if (iconGravity == ICON_GRAVITY_END || iconGravity == ICON_GRAVITY_TEXT_END) {
			setCompoundDrawablesRelative(null, null, icon, null);
		}
	}
	
	@Override
	protected void onSizeChanged(int w, int h, int oldw, int oldh) {
		super.onSizeChanged(w, h, oldw, oldh);
		updateIconPosition(w, h);
	}
	
	@Override
	protected void onTextChanged(CharSequence text, int start, int lengthBefore, int lengthAfter) {
		super.onTextChanged(text, start, lengthBefore, lengthAfter);
		updateIconPosition(getMeasuredWidth(), getMeasuredHeight());
	}
	
	private void updateIconPosition(int buttonWidth, int buttonHeight) {
		if (icon == null || getLayout() == null) return;
		
		if (iconGravity == ICON_GRAVITY_TEXT_START || iconGravity == ICON_GRAVITY_TEXT_END) {
			int localIconSize = iconSize == 0 ? icon.getIntrinsicWidth() : iconSize;
			
			int paddingStart = Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 ? getPaddingStart() : getPaddingLeft();
			int paddingEnd = Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 ? getPaddingEnd() : getPaddingRight();
			
			int newIconLeft = (buttonWidth - getTextWidth() - paddingEnd - localIconSize - (int) iconPadding - paddingStart) / 2;
			
			boolean isRTL = Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 && getLayoutDirection() == View.LAYOUT_DIRECTION_RTL;
			
			if (isRTL != (iconGravity == ICON_GRAVITY_TEXT_END)) {
				newIconLeft = -newIconLeft;
			}
			
			newIconLeft = Math.max(0, newIconLeft);
			
			if (iconLeft != newIconLeft) {
				iconLeft = newIconLeft;
				updateIcon(false);
			}
		} else {
			if (iconLeft != 0) {
				iconLeft = 0;
				updateIcon(false);
			}
		}
	}
	
	private int getTextWidth() {
		Paint textPaint = getPaint();
		String text = getText().toString();
		if (getTransformationMethod() != null) {
			text = getTransformationMethod().getTransformation(text, this).toString();
		}
		if (getLayout() != null) {
			return Math.min((int) textPaint.measureText(text), getLayout().getEllipsizedWidth());
		} else {
			return (int) textPaint.measureText(text);
		}
	}
	
	public void setIcon(@Nullable Drawable drawable) {
		if (this.icon != drawable) {
			this.icon = drawable;
			updateIcon(true);
			updateIconPosition(getMeasuredWidth(), getMeasuredHeight());
		}
	}
	
	public void setIcon(int resId) {
		if (resId != 0) {
			Drawable drawable;
			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
				drawable = getContext().getDrawable(resId);
			} else {
				drawable = getContext().getResources().getDrawable(resId);
			}
			setIcon(drawable);
		} else {
			setIcon((Drawable) null);
		}
	}
	
	public void setCornerRadius(float radius) {
		this.cornerRadius = radius;
		applyBackground();
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
			invalidateOutline();
		}
	}
	
	public void setIconTint(int color) {
		this.iconTint = color;
		updateIcon(false);
	}
	
	@Override
	public void setBackgroundColor(int color) {
		this.bgColor = color;
		applyBackground();
	}
	
	public void setIconSize(int sizePx) {
		this.iconSize = sizePx;
		updateIcon(true);
	}
	
	public void setRippleColor(int color) {
		this.rippleColor = color;
		applyBackground();
	}
	
	public void setStroke(float width, int color) {
		this.strokeWidth = width;
		this.strokeColor = color;
		applyBackground();
	}
	
	public void setStrokeColor(int color) {
		this.strokeColor = color;
		applyBackground();
	}
	
	public void setStrokeWidth(float width) {
		this.strokeWidth = width;
		applyBackground();
	}
	
	public void setIconGravity(int gravity) {
		if (this.iconGravity != gravity) {
			this.iconGravity = gravity;
			updateIconPosition(getMeasuredWidth(), getMeasuredHeight());
		}
	}
	
	private int dpToPx(Context context, float dp) {
		return (int) (dp * context.getResources().getDisplayMetrics().density + 0.5f);
	}
}

File: app/src/main/java/com/aexon/material/button/AexonButton.kt
package com.aexon.material.button

import android.content.Context
import android.content.res.ColorStateList
import android.content.res.TypedArray
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.RippleDrawable
import android.os.Build
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.widget.Button
import com.aexon.R
import com.aexon.annotation.NonNull
import com.aexon.annotation.Nullable
import com.aexon.core.AexonMath
import kotlin.math.max
import kotlin.math.min

class AexonButton : Button {
	
	companion object {
		const val ICON_GRAVITY_START = 0x1
		const val ICON_GRAVITY_TEXT_START = 0x2
		const val ICON_GRAVITY_END = 0x3
		const val ICON_GRAVITY_TEXT_END = 0x4
	}
	
	private var cornerRadius = 0f
	private var iconGravity = ICON_GRAVITY_START
	private var iconPadding = 0f
	private var iconTint = Integer.MIN_VALUE
	private var iconSize = 0
	private var iconLeft = 0
	private var icon: Drawable? = null
	private var bgColor = Color.TRANSPARENT
	private var rippleColor = Color.argb(40, 255, 255, 255)
	private var strokeColor = Color.TRANSPARENT
	private var strokeWidth = 0f
	
	constructor(@NonNull context: Context) : super(context) {
		init(context, null)
	}
	
	constructor(@NonNull context: Context, @Nullable attrs: AttributeSet?) : super(context, attrs) {
		init(context, attrs)
	}
	
	constructor(@NonNull context: Context, @Nullable attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
		init(context, attrs)
	}
	
	private fun init(@NonNull context: Context, @Nullable attrs: AttributeSet?) {
		isAllCaps = false
		minimumWidth = AexonMath.dpToPxInt(context, 88f)
		minimumHeight = AexonMath.dpToPxInt(context, 36f)
		includeFontPadding = false
		gravity = Gravity.CENTER
		
		if (attrs != null) {
			val a = context.obtainStyledAttributes(attrs, R.styleable.AexonButton)
			cornerRadius = a.getDimension(R.styleable.AexonButton_cornerRadius, 0f)
			iconGravity = a.getInt(R.styleable.AexonButton_iconGravity, ICON_GRAVITY_START)
			iconPadding = a.getDimension(R.styleable.AexonButton_iconPadding, AexonMath.dpToPxInt(context, 8f).toFloat())
			iconTint = a.getColor(R.styleable.AexonButton_iconTint, Integer.MIN_VALUE)
			iconSize = a.getDimension(R.styleable.AexonButton_iconSize, 0f).toInt()
			icon = a.getDrawable(R.styleable.AexonButton_icon)
			bgColor = a.getColor(R.styleable.AexonButton_buttonColor, Color.TRANSPARENT)
			rippleColor = a.getColor(R.styleable.AexonButton_rippleColor, Color.argb(40, 255, 255, 255))
			strokeColor = a.getColor(R.styleable.AexonButton_strokeColor, Color.TRANSPARENT)
			strokeWidth = a.getDimension(R.styleable.AexonButton_strokeWidth, 0f)
			a.recycle()
		}
		
		if (bgColor == Color.TRANSPARENT) {
			val bg = background
			if (bg is android.graphics.drawable.ColorDrawable) {
				bgColor = bg.color
			}
		}
		
		compoundDrawablePadding = iconPadding.toInt()
		applyBackground()
		updateIcon(true)
	}
	
	private fun applyBackground() {
		val shape = GradientDrawable().apply {
			setColor(bgColor)
			setCornerRadius(cornerRadius)
			
			if (strokeWidth > 0 && strokeColor != Color.TRANSPARENT) {
				setStroke(strokeWidth.toInt(), strokeColor)
			}
		}
		
		val ripple = RippleDrawable(ColorStateList.valueOf(rippleColor), shape, null)
		background = ripple
		
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
			outlineProvider = object : android.view.ViewOutlineProvider() {
				override fun getOutline(view: View, outline: android.graphics.Outline) {
					if (view.width > 0 && view.height > 0) {
						outline.setRoundRect(0, 0, view.width, view.height, cornerRadius)
					}
				}
			}
			clipToOutline = true
			invalidateOutline()
		}
	}
	
	private fun updateIcon(needsIconReset: Boolean) {
		if (icon != null) {
			icon = icon!!.mutate()
			
			val width = if (iconSize != 0) iconSize else icon!!.intrinsicWidth
			val height = if (iconSize != 0) iconSize else icon!!.intrinsicHeight
			icon!!.setBounds(iconLeft, 0, iconLeft + width, height)
			
			if (iconTint != Integer.MIN_VALUE) {
				icon!!.setColorFilter(iconTint, PorterDuff.Mode.SRC_IN)
			} else {
				icon!!.clearColorFilter()
			}
		}
		
		if (needsIconReset) {
			resetIconDrawable()
		}
	}
	
	private fun resetIconDrawable() {
		when (iconGravity) {
			ICON_GRAVITY_START, ICON_GRAVITY_TEXT_START -> setCompoundDrawablesRelative(icon, null, null, null)
			ICON_GRAVITY_END, ICON_GRAVITY_TEXT_END -> setCompoundDrawablesRelative(null, null, icon, null)
		}
	}
	
	override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
		super.onSizeChanged(w, h, oldw, oldh)
		updateIconPosition(w, h)
	}
	
	override fun onTextChanged(text: CharSequence?, start: Int, before: Int, after: Int) {
		super.onTextChanged(text, start, before, after)
		updateIconPosition(measuredWidth, measuredHeight)
	}
	
	private fun updateIconPosition(buttonWidth: Int, buttonHeight: Int) {
		if (icon == null || layout == null) return
		
		if (iconGravity == ICON_GRAVITY_TEXT_START || iconGravity == ICON_GRAVITY_TEXT_END) {
			val localIconSize = if (iconSize == 0) icon!!.intrinsicWidth else iconSize
			
			val paddingStart = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) paddingStart else paddingLeft
			val paddingEnd = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) paddingEnd else paddingRight
			
			var newIconLeft = (buttonWidth - textWidth - paddingEnd - localIconSize - iconPadding.toInt() - paddingStart) / 2
			
			val isRTL = Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 && layoutDirection == View.LAYOUT_DIRECTION_RTL
			
			if (isRTL != (iconGravity == ICON_GRAVITY_TEXT_END)) {
				newIconLeft = -newIconLeft
			}
			
			newIconLeft = max(0, newIconLeft)
			
			if (iconLeft != newIconLeft) {
				iconLeft = newIconLeft
				updateIcon(false)
			}
		} else {
			if (iconLeft != 0) {
				iconLeft = 0
				updateIcon(false)
			}
		}
	}
	
	private val textWidth: Int
	get() {
		val textPaint = paint
		var text = getText().toString()
		if (transformationMethod != null) {
			text = transformationMethod!!.getTransformation(text, this).toString()
		}
		return if (layout != null) {
			min(textPaint.measureText(text).toInt(), layout!!.ellipsizedWidth)
		} else {
			textPaint.measureText(text).toInt()
		}
	}
	
	fun setIcon(@Nullable drawable: Drawable?) {
		if (this.icon != drawable) {
			this.icon = drawable
			updateIcon(true)
			updateIconPosition(measuredWidth, measuredHeight)
		}
	}
	
	fun setIcon(resId: Int) {
		if (resId != 0) {
			val drawable = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
				context.getDrawable(resId)
			} else {
				context.resources.getDrawable(resId)
			}
			setIcon(drawable)
		} else {
			setIcon(null as Drawable?)
		}
	}
	
	fun setCornerRadius(radius: Float) {
		this.cornerRadius = radius
		applyBackground()
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
			invalidateOutline() 
		}
	}
	
	fun setIconTint(color: Int) {
		this.iconTint = color
		updateIcon(false)
	}
	
	override fun setBackgroundColor(color: Int) {
		this.bgColor = color
		applyBackground()
	}
	
	fun setIconSize(sizePx: Int) {
		this.iconSize = sizePx
		updateIcon(true)
	}
	
	fun setRippleColor(color: Int) {
		this.rippleColor = color
		applyBackground()
	}
	
	fun setStroke(width: Float, color: Int) {
		this.strokeWidth = width
		this.strokeColor = color
		applyBackground()
	}
	
	fun setStrokeColor(color: Int) {
		this.strokeColor = color
		applyBackground()
	}
	
	fun setStrokeWidth(width: Float) {
		this.strokeWidth = width
		applyBackground()
	}
	
	fun setIconGravity(gravity: Int) {
		if (this.iconGravity != gravity) {
			this.iconGravity = gravity
			updateIconPosition(measuredWidth, measuredHeight)
		}
	}
}